rust/etw_helpers/src/processtrace.rs

```diff
@@ -0,0 +1,245 @@
+                    as *const c_void as *mut c_void;
+
+                props
+            };
+
+            let hndl = OpenTraceA(&mut log.props);
+            if hndl.0 == u64::MAX {
+                let err = GetLastError();
+                Err(err.into())
+            } else {
+                *inner.hndl.lock().unwrap() = Some(ProcessTraceHandleWrapper(hndl));
+                Ok(ProcessTraceHandle { inner, ctx: clone })
+            }
+        }
+    }
+
+    unsafe extern "system" fn event_record_callback(event_record: *mut EVENT_RECORD) {
+        let ctx = (*event_record).UserContext as *mut InnerProcessTraceHandle<C>;
+        if !ctx.is_null() {
+            // It's not safe to let a panic cross back into C code.
+            // Use AssertUnwindSafe because we will always abort in the event of a panic.
+            let err = catch_unwind(AssertUnwindSafe(|| {
+                let result = (*ctx).inner_callback(event_record);
+                if result.is_err() {
+                    (*ctx).close_trace(false);
+                }
+            }));
+            if err.is_err() {
+                std::process::abort();
+            }
+        }
+    }
+
+    pub fn process_trace(self) -> Result<ProcessTraceThread<C>, windows::core::Error> {
+        let thread = spawn_process_trace_thread(self.inner.clone(), self.ctx);
+
+        Ok(ProcessTraceThread {
+            thread: Some(thread),
+            inner: self.inner,
+        })
+    }
+}
+
+fn spawn_process_trace_thread<C>(
+    inner: Arc<InnerProcessTraceHandle<C>>,
+    mut ctx: ManuallyDrop<Arc<InnerProcessTraceHandle<C>>>,
+) -> JoinHandle<Result<(), windows::core::Error>>
+where
+    C: EventConsumer + Send + Sync + 'static,
+{
+    let handles = [***inner.hndl.lock().unwrap().as_ref().as_ref().unwrap()];
+    unsafe {
+        std::thread::spawn(move || {
+            let err = ProcessTrace(&handles, None, None);
+            inner.process_trace_complete(err.into());
+            ManuallyDrop::drop(&mut ctx);
+            if err.is_err() {
+                Err(windows::core::Error::from(err))
+            } else {
+                Ok(())
+            }
+        })
+    }
+}
+
+pub struct ProcessTraceThread<C>
+where
+    C: EventConsumer + Send + Sync + 'static,
+{
+    thread: Option<JoinHandle<Result<(), windows::core::Error>>>,
+    inner: Arc<InnerProcessTraceHandle<C>>,
+}
+
+impl<C> ProcessTraceThread<C>
+where
+    C: EventConsumer + Send + Sync,
+{
+    pub fn stop_and_wait(&mut self) -> Result<(), windows::core::Error> {
+        let thread = self.stop_and_get_thread();
+        thread.join().unwrap()
+    }
+
+    pub fn stop_and_get_thread(&mut self) -> JoinHandle<Result<(), windows::core::Error>> {
+        self.inner.close_trace(false);
+        self.thread.take().unwrap()
+    }
+}
+
+pub trait EventConsumer {
+    unsafe fn on_event_raw(&self, evt: *mut EVENT_RECORD) -> Result<(), windows::core::Error>;
```

Comment:
This is always an option for people that don't like channels, but it's their burden to not return until all usages of the pointer are dropped (hence the unsafe classification).

Reply:
If the signature were changed to `evt: &EVENT_RECORD` then it would be safe. (This would also force the implementation to be more careful with lifetimes, which would be annoying to them, but it can be made to work.)